### PR TITLE
Return error messages when state transitions fail

### PR DESCRIPTION
**Pull request by tijmenb**
### Context

We have added a state machine in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/278. We currently raise an error and return `500` when an API user tries to do an impossible thing.

### Changes proposed in this pull request

- Return a specific error from the decision endpoints
- Remove the tests for the individual decisions - they are testing the same thing as the integration tests. Talked about this with @RobertM905. 

### Link to Trello card

https://trello.com/c/u06H8NSO

### CODE REVIEW

File: app/controllers/vendor_api/confirm_candidate_enrolment_controller.rb [hunk 1]
```diff
@@ -9,12 +9,14 @@ def confirm
...
 
       application_choice = ApplicationChoice.find(params[:application_id])
-      result = ConfirmEnrolment.new(application_choice: application_choice).call
+      confirmation = ConfirmEnrolment.new(application_choice: application_choice)
 
-      if result.successful?
+      if confirmation.save
         render json: {
-          data: SingleApplicationPresenter.new(result.application_choice).as_json,
+          data: SingleApplicationPresenter.new(application_choice).as_json,
```
Comment (duncanjbrown) [inline app/controllers/vendor_api/confirm_candidate_enrolment_controller.rb]: Should we `.reload` here (and in the other controllers)? the state that was loaded at line 11 will hopefully not be the same as the state after update

File: spec/requests/vendor_api/reject_application_spec.rb [hunk 2]
```diff
@@ -25,8 +25,17 @@
...
+
+    post "/api/v1/applications/#{application_choice.id}/reject", params: {}
+
+    expect(response).to have_http_status(422)
+    expect(parsed_response).to be_valid_against_openapi_schema('BadRequestBodyResponse')
+  end
+
   it 'returns not found error when the application was not found' do
-    post '/api/v1/applications/non-existent-id/confirm-enrolment'
+    post '/api/v1/applications/non-existent-id/reject'
```
Comment (RobertM905) [inline spec/requests/vendor_api/reject_application_spec.rb]: Nice catch! 👍

File: app/services/confirm_enrolment.rb [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
   end
 
-  def call
-    ApplicationStateChange.new(@application_choice).confirm_enrolment!
-
-    Response.new(true, @application_choice)
+  def save
+    ApplicationStateChange.new(application_choice).confirm_enrolment!
+    true
+  rescue Workflow::NoTransitionAllowed => e
```
Comment (duncanjbrown) [inline app/services/confirm_enrolment.rb]: Ought we to include `ActiveModel::Validations` in these objects? Then we get error key-value matching for free, plus `.full_sentences` etc.

File: app/services/confirm_enrolment.rb [hunk 4]
```diff
@@ -1,13 +1,19 @@
 class ConfirmEnrolment
-  Response = Struct.new(:successful?, :application_choice)
+  attr_reader :errors, :application_choice
```
Comment (duncanjbrown) [inline app/services/confirm_enrolment.rb]: I think we can 🔪 `:application_choice` here if we're hanging on to the reference  in the controller?

File: app/controllers/vendor_api/confirm_candidate_enrolment_controller.rb (hunk 1 above)
Comment (tijmenb) [reply app/controllers/vendor_api/confirm_candidate_enrolment_controller.rb]: I'm assuming the object is updated together with the database. As long as the test still passes I don't tink we have to do anything more:

https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/b3e547734a18028b7b98d67b2b2fa91505a3f161/spec/requests/vendor_api/confirm_enrolment_spec.rb#L18-L27

File: spec/requests/vendor_api/reject_application_spec.rb (hunk 2 above)
Comment (chubberlisk) [reply spec/requests/vendor_api/reject_application_spec.rb]: Just a lil fyi incase, Duncan has created a PR to fix this, see https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/288.

File: app/controllers/vendor_api/decisions_controller.rb [hunk 5]
```diff
@@ -59,5 +55,20 @@ def invalid_metadata(_e)
...
+      if decision.save
+        render json: { data: SingleApplicationPresenter.new(application_choice).as_json }
+      else
+        render_error_response(decision.errors)
+      end
+    end
+
+    # Takes a object with ActiveModel::Validations and render the `errors`
+    # as API response.
+    def render_error_response(errors)
```
Comment (alexdesi) [inline app/controllers/vendor_api/decisions_controller.rb]: this is just for '422', should this be clear from the method name?
`render_422_error_response` or
`render_bad_request_error_response`

File: app/controllers/vendor_api/decisions_controller.rb [hunk 6]
```diff
@@ -6,51 +6,47 @@ class DecisionsController < VendorApiController
...
+      decision = MakeAnOffer.new(
+        application_choice: application_choice,
+        offer_conditions: params[:data],
+      )
 
-      make_an_offer = MakeAnOffer.new(application_choice: application_choice, offer_conditions: params[:data]).call
-      if make_an_offer.successful?
-        render json: { data: SingleApplicationPresenter.new(make_an_offer.application_choice).as_json }
-      end
+      repond_to_decision(decision)
```
Comment (duncanjbrown) [inline app/controllers/vendor_api/decisions_controller.rb]: :+1:

File: app/controllers/vendor_api/decisions_controller.rb [hunk 7]
```diff
@@ -59,5 +55,20 @@ def invalid_metadata(_e)
...
+        render json: { data: SingleApplicationPresenter.new(application_choice).as_json }
+      else
+        render_error_response(decision.errors)
+      end
+    end
+
+    # Takes a object with ActiveModel::Validations and render the `errors`
+    # as API response.
+    def render_error_response(errors)
+      errors = errors.map { |_key, message| { error: 'BadRequest', message: message } }
```
Comment (alexdesi) [inline app/controllers/vendor_api/decisions_controller.rb]: not sure we should reassign the input param`errors`,
should we rename it to `error_messages` (or something)?

File: app/services/reject_application.rb [hunk 8]
```diff
@@ -1,17 +1,20 @@
...
+  include ActiveModel::Validations
 
   def initialize(application_choice:, rejection:)
     @application_choice = application_choice
     @rejection = rejection
   end
 
-  def call
+  def save
     ApplicationStateChange.new(@application_choice).reject_application!
```
Comment (duncanjbrown) [inline app/services/reject_application.rb]: Should this and the below update be wrapped in a transaction?

File: spec/requests/vendor_api/confirm_enrolment_spec.rb [hunk 9]
```diff
@@ -56,4 +56,14 @@
...
     expect(error_response['message']).to eql('Could not find an application with ID non-existent-id')
   end
+
+  it 'returns an error when trying to transition to an invalid state' do
+    application_choice = create(:application_choice, status: 'rejected')
+
+    post "/api/v1/applications/#{application_choice.id}/confirm-enrolment",
+         params: { meta: valid_metadata }
+
+    expect(response).to have_http_status(422)
```
Comment (duncanjbrown) [inline spec/requests/vendor_api/confirm_enrolment_spec.rb]: How about https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409?

Comment (tijmenb) [reply spec/requests/vendor_api/confirm_enrolment_spec.rb]: I think the code will be used for all decision errors, where 409 might not be applicable.

File: app/controllers/vendor_api/decisions_controller.rb (hunk 5 above)
Comment (tijmenb) [reply app/controllers/vendor_api/decisions_controller.rb]: Done

File: app/services/reject_application.rb (hunk 8 above)
Comment (tijmenb) [reply app/services/reject_application.rb]: Done

**Pull request merged**